src/main/kotlin/platform/mixin/expression/gui/FlowStrings.kt

```diff
@@ -0,0 +1,205 @@
+        LocalVariables.getLocalVariableAt(project, clazz, method, location, index)
+    }?.name ?: "<local $index>"
+    val isStore = flow.insn.opcode in Opcodes.ISTORE..Opcodes.ASTORE
+    return localName + if (isStore) " =" else ""
+}
+
+private fun fieldString(insn: FieldInsnNode): String {
+    val isStatic = insn.opcode in Opcodes.GETSTATIC..Opcodes.PUTSTATIC
+    val isWrite = insn.opcode == Opcodes.PUTFIELD || insn.opcode == Opcodes.PUTSTATIC
+    val owner = if (isStatic) shortOwner(insn.owner) else ""
+    return "$owner.${insn.name}" + if (isWrite) " =" else ""
+}
+
+private fun methodString(insn: MethodInsnNode, methodCallType: MethodCallType): String {
+    val owner = when (methodCallType) {
+        MethodCallType.NORMAL -> ""
+        MethodCallType.SUPER -> "super"
+        MethodCallType.STATIC -> shortOwner(insn.owner)
+    }
+    return "$owner.${insn.name}${shortParams(insn.desc)}"
+}
+
+private fun castString(type: Type): String = "(${type.shortName})"
+
+private fun complexCmpString(opcode: Int, jumpOpcode: Int): String {
+    val isG = opcode == Opcodes.FCMPG || opcode == Opcodes.DCMPG
+    return when (jumpOpcode) {
+        Opcodes.IFEQ, Opcodes.IFNE -> "== or !="
+        Opcodes.IFLT, Opcodes.IFGE -> if (isG) "<" else ">="
+        Opcodes.IFGT, Opcodes.IFLE -> if (isG) "<=" else ">"
+        else -> "Unknown jump"
+    }
+}
+
+private fun newArrayString(flow: FlowValue): String? {
+    val insn = flow.insn
+    val type = when (insn.opcode) {
+        Opcodes.NEWARRAY, Opcodes.ANEWARRAY -> ExpressionASMUtils.getUnaryType(insn)
+        Opcodes.MULTIANEWARRAY -> ExpressionASMUtils.getNaryType(insn)
+        else -> return null
+    }
+    val prefix = "new ${type.shortName}"
+    if (!flow.hasDecoration(FlowDecorations.ARRAY_CREATION_INFO)) {
+        return prefix
+    }
+    return "$prefix{ }"
+}
+
+private fun opcodeString(opcode: Int) = when (opcode) {
+    Opcodes.ATHROW -> "throw"
+    in Opcodes.IRETURN..Opcodes.RETURN -> "return"
+    in Opcodes.IADD..Opcodes.DADD -> "+"
+    in Opcodes.ISUB..Opcodes.DSUB -> "-"
+    in Opcodes.IMUL..Opcodes.DMUL -> "*"
+    in Opcodes.IDIV..Opcodes.DDIV -> "/"
+    in Opcodes.IREM..Opcodes.DREM -> "%"
+    in Opcodes.INEG..Opcodes.DNEG -> "-"
+    Opcodes.ISHL, Opcodes.LSHL -> "<<"
+    Opcodes.ISHR, Opcodes.LSHR -> ">>"
+    Opcodes.IUSHR, Opcodes.LUSHR -> ">>>"
+    Opcodes.IAND, Opcodes.LAND -> "&"
+    Opcodes.IOR, Opcodes.LOR -> "|"
+    Opcodes.IXOR, Opcodes.LXOR -> "^"
+    Opcodes.IF_ACMPEQ, Opcodes.IF_ICMPEQ, Opcodes.IF_ACMPNE, Opcodes.IF_ICMPNE -> "== or !="
+    Opcodes.IF_ICMPLT, Opcodes.IF_ICMPGE -> "< or >="
+    Opcodes.IF_ICMPLE, Opcodes.IF_ICMPGT -> "<= or >"
+    Opcodes.ARRAYLENGTH -> ".length"
+    in Opcodes.IALOAD..Opcodes.SALOAD -> "[]"
+    in Opcodes.IASTORE..Opcodes.SASTORE -> "[] ="
+    else -> null
+}
+
+private fun String.escape() =
+    StringUtil.escapeStringCharacters(this)
+        .replace("'", "\\\\'")
```

Comment:
I'm pretty sure there's a way to do this directly by changing what the quotation character is. Check what `escapeStringCharacters` delegates to. There is also `escapeCharCharacters` but I'm not sure that will work for multi-character strings

Reply:
We need to escape both double quotes and single quotes, single quotes in a special way, Id be surprised if there's a built in method for that

Reply:
Interesting, I didn't think we'd need to escape double quotes.

Reply:
The whole expression is inside a string literal and this is meant to match what you'd type there